dnprogs/apps/dnmount.8
.TH DNMOUNT 1 "Aug 28 1998" "DECnet utilities"

.SH NAME
dnmount \- Mounts DEC/VMS directories as Linux Filesystems

.SH SYNOPSIS
.B dnmount
[options] VMS_DIRECTORY mount_point
.br
Options:
.br
[\-h] [\-u uid] [\-g gid] 

.br
-h prints a help screen
.br
-u specifies UID for mounted filesystem.
.br
-g specifies GID for mounted filesystem.

.SH DESCRIPTION
.PP
dnmount mounts VMS directories structures as Linux Filesystems.
.br
 
.br
The VMS_DIRECTORY should be specified in the usual transparent DECnet format
of node"username password"::[directory] or node.
.br

.SH EXAMPLES
.br

.br
.PP
    dnmount -u serrat 'mv3100::sys$manager:' /vax

.PP
    dnmount  'mv3100"serrat passwrd"::sys$login:' /vaxuser
.br


.PP
    dnmount  'mv3100"system passwrd"::dka0:[sys0]' /vax

.br
.SH HINTS
.br
    Use umount to dismount a dapfs mounted with dnmount.

.SH AUTHOR 
.br
    Eduardo Marcelo Serrat <emserrat@geocities.com>


dnprogs/NEWS
dnprogs (2.61) unstable; urgency=low

  * Include libvaxdata/src/test.c in the distribution tarball
  * Fix password length in DAP connections. It was incorrectly set at 12 
    rather than 40
 
 -- Chrissie Caulfield <chrissie@debian.org>  Sun, 31 Mar 2013 11:18:32 +0000

dnprogs (2.60+x32) unreleased; urgency=low

  * Switch from using deprecated sysctl(2) to writing to /proc/sys.  This
    fixes a build failure on x32.

 -- Daniel Schepler <schepler@debian.org>  Tue, 12 Feb 2013 12:21:31 -0800

dnprogs (2.60) unstable; urgency=low

  * Fixed lots of Debian things that caused lintian problems and 
    upgraded policy version
    Closes: #673408
  * Fix some warnings that might be bugs
  * Incorporate NMU into main tree.

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 26 May 2012 09:54:41 +0100

dnprogs (2.59+nmu2) unstable; urgency=medium

  * NMU
  * dnprogs binaries are Linux-only, so state that. Move dnet-common back
    to binary-all. Closes: #665886

 -- Steve McIntyre <93sam@debian.org>  Mon, 21 May 2012 17:43:14 +0100

dnprogs (2.59+nmu1) unstable; urgency=low

  * NMU
  * Reduce libdnet's recommends on dnet-common to suggests again.
    Closes: #655740. Adapted patch from Jonathan Nieder.

 -- Steve McIntyre <93sam@debian.org>  Fri, 18 May 2012 13:29:53 +0100

dnprogs (2.59) unstable; urgency=low

  * Changed package dnet-common from arch all to any and
    updated dependecies. (Closes: #665886)

 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 27 Apr 2012 09:30:19 +0100

dnprogs (2.58) unstable; urgency=medium

  * Fixed lintian warnings:
    - Fixed changelog entry of last release to have correct line length.
    - Corrected debconf dependency in dnet-common.
    - Corrected versioned dependency in dnprogs on dnet-common.
    - Updated existing overrides to match new syntax.
  * Updated linkage of libdnet_daemon and libdap. (Closes: #646353)
  * Fixed compile warning in dndir when using -Werror=format-security
    (Closes: #646613)
  * Fixed typo in debian/dnet-common.config
  * Do not try to install files into packages on FreeBSD which are
    not build on FreeBSD.
  * Link dapfs with libpthread for new libfuse compatibility
  * Updated nl.po (Closes: #652349)
  * Added pl.po (Closes: #661894)
  * Added it.po (Closes: #663135)
  * Change module-init-tools dependency to kmod (Closes: #662684)

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 10 Mar 2012 10:32:08 +0000

dnprogs (2.57) unstable; urgency=medium

  [ Chrissie Caulfield ]
  * Completed removal of modutils configuration file code
  * Fix presence of CVS directories in tarball
  * updated 'clean' target

  [ Philipp Schafft ]
  * Use /dev/pts PTYs on 64 bit systems (fix makefile bug)
  * Updated translations Debconf templates:
    - German translation (Closes: #615268)
    - French translation (Closes: #615039)
    - Danish translation (Closes: #633984) (added)
    - Slovak translation (Closes: #639629) (added)
  * Look in yet another place for libcrypt (better fix than from NMUs)
  * Added a comment about the detecting of libcrypt. (See: #629664)
  * "ethernet" was capitalized in extended description (Closes: #636798)
  * Fixed some lintian warnings (mainly new rules targets).
  * dnet-common: Do not configure DECnet unless the user requested it.
    (Removed the bad workaround done by NMU) (See: #637179)
  * dnet-common: Added preinst script to clean up the broken config files
    from the last NMU. (Closes: #640861)
  * Upgraded libdnet's suggest to dnet-common to recommend again
    (Closes: #608807)

 -- Chrissie Caulfield <chrissie@debian.org>  Thu,  8 Sep 2011 16:05:47 +0200

dnprogs (2.56.1+nmu1) unstable; urgency=low

  * Non-maintainer upload.
  * libdnet: Only suggest dnet-common package (Closes: #608807)
  * dnet-common: Do not change MAC addresses unless user chooses to
    configure DECnet (Closes: #637179)
  * dnet-common: Restore ability to configure DECnet name and address
    through debconf
  * dnet-common: Remove obsolete modutils configuration file

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 25 Aug 2011 14:26:12 +0100

dnprogs (2.56.1) unstable; urgency=low

  * Non-maintainer upload.
  * Add patch from Ubuntu to fix FTBFS with libcrypt moved to the
    multiarch path.

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 09 Aug 2011 11:58:06 +0200


dnprogs (2.56) unstable; urgency=low

  * Search for libgcrypt in 64bit library paths, too. (thanks to Kamil)
  * Package dnet-common now asks if network interfaces should be configrued
    for DECnet (Closes: #608807)
    Updated translations Debconf templates:
    - Russian translation (Closes: #610642)
    - Portuguese translation (Closes: #611034)
    - Swedish translation (Closes: #611296)
    - Czech translation (Closes: #611430)
    - Japanese translation (Closes: #611448)
    - Spanish translation (Closes: #611547)
  * Build against version 25 of FUSE
    Closes: #598831
  * Fixed some lintian warnings (typos and such).
  * Raised priority of package libdnet from extra to optional
    to comply policy manual section 2.5 "Priorities".

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 22 Feb 2011 10:38:23 +0000

dnprogs (2.55) unstable; urgency=low

  * Build against version 25 of FUSE
    Closes: #598831

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 5 Oct 2010 09:23:12 +0100

dnprogs (2.54) unstable; urgency=low

  * Don't build libvaxdata test program by default.
    Closes: #598752
 
 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 1 Oct 2010 20:44:09 +0100

dnprogs (2.53) unstable; urgency=low

  * Done some more porting to FreeBSD, corrected Build-Depends for hurd-i386
    Closes: #588185
  * Added new tool node(1) - Node name lookup tool
  * Added support to libdnet_daemon to read /etc/nodes.{allow,deny}
  * Fixed DoS bug in dnet_daemon()
    Closes: #592015 
  * Update libvaxdata to version 1.2, fixes double conversions

 -- Christine Caulfield <Christine.Caulfield@googlemail.com>  Thu, 19 Jul 2010 10:19:32 +0100


dnprogs/apps/dnmount.c
/* dnmount.c */
/******************************************************************************
    (c) 1995-1998 E.M. Serrat             emserrat@geocities.com

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
*******************************************************************************/

#include <stdio.h>
#include <errno.h>
#include <ctype.h>
#include <mntent.h>
#include <pwd.h>
#include <grp.h>
#include <unistd.h>
#include <signal.h>
#include <string.h>
#include <fcntl.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <sys/mount.h>
#include <netdnet/dn.h>
#include <netdnet/dnetdb.h>
#include <asm/types.h>
#ifndef BLOCK_SIZE
#include <linux/fs.h>
#endif
#include <linux/dap_fs.h>

/* DECnet phase IV limits */
#define MAX_NODE      6
#define MAX_USER     12
#define MAX_PASSWORD 12
#define MAX_ACCOUNT  12
#ifndef FALSE
#define TRUE 1
#define FALSE 0
#endif

struct	sockaddr_dn			sockaddr;
struct	accessdata_dn			accessdata;
struct	nodeent				*dp;
struct	dapfs_mount_data		data;
char  					node[MAX_NODE+1],dirname[250],
					vms_directory[250],mount_point[80],
					*lasterror;
/*-----------------------------------------------------------------------*/
void
usage(void)
{
	printf("\nusage: dnmount [OPTIONS] vms_directory mount_point\n");
	printf("OPTIONS:\n");
	printf("-u uid		User  ID for mounted filesystem\n");
	printf("-g gid		Group ID for mounted filesystem\n");
	printf("-h -?		Display this help\n\n");
}
/*-----------------------------------------------------------------------*/
/* Christine's Parser for VMS file specifications
 * It was taken from dndir.c
 *
*/
/*-------------------------------------------------------------------------*/
/* Parse the filename into its component parts */
static int parse(char *fname)
{
    enum  {NODE, USER, PASSWORD, ACCOUNT, NAME, FINISHED} state;
    int   n0=0;
    int   n1=0;
    char  sep, term;
    int   i;
    char *local_user;

    state = NODE; /* Node is mandatory */

    while (state != FINISHED)
    {
	switch (state)
	{
	case NODE:
            if (fname[n0] != ':' && fname[n0] != '\"' &&
                fname[n0] != '\'' && fname[n0] != '/')
            {
                if (n1 >= MAX_NODE ||
                    fname[n0] == ' ' || fname[n0] == '\n')
                {
                    lasterror = (char *)"File name parse error";
                    return FALSE;
                }
                node[n1++] = fname[n0++];
            }
            else
            {
                node[n1] = '\0';
                n1 = 0;
                switch (fname[n0])
                {
                case '\"':
                case '\'':
                    sep = ' ';
                    term = fname[n0++];
                    state = USER;
                    break;

                case '/':
                    sep = '/';
                    term = 0;
                    n0++;
                    state = USER;
                    break;

                default:
                    n0 += 2;
                    state = NAME;
                    break;
                }
            }
            break;

	case USER:
            if (fname[n0] != sep && fname[n0] != (term ? term : ':'))
            {
                if (n1 >= MAX_USER)
                {
                    lasterror = (char *)"File name parse error";
                    return FALSE;
                }
                accessdata.acc_user[n1++] = fname[n0++];
            }
            else
            {
                accessdata.acc_user[n1] = '\0';
                n1 = 0;
                if (fname[n0] == sep)
                {
                    state = PASSWORD;
                    n0++;
                }
                else
                {
                    state = NAME;
                    n0 += term ? 3 : 2;
                }
            }
            break;

	case PASSWORD:
            if (fname[n0] != sep && fname[n0] != (term ? term : ':'))
            {
                if (n1 >= MAX_PASSWORD)
                {
                    lasterror = (char *)"File name parse error";
                    return FALSE;
                }
                accessdata.acc_pass[n1++] = fname[n0++];
            }
            else
            {
                accessdata.acc_pass[n1] = '\0';
                n1 = 0;
                if (fname[n0] == sep)
                {
                    state = ACCOUNT;
                    n0++;
                }
                else
                {
                    state = NAME;
                    n0 += term ? 3 : 2;
                }
            }
            break;

	case ACCOUNT:
            if (fname[n0] != sep && fname[n0] != (term ? term : ':'))
            {
                if (n1 >= MAX_ACCOUNT)
                {
                    lasterror = (char *)"File name parse error";
                    return FALSE;
                }
                accessdata.acc_acc[n1++] = fname[n0++];
            }
            else
            {
                accessdata.acc_acc[n1] = '\0';
                n1 = 0;
                state = NAME;
                n0 += term ? 3 : 2;
            }
            break;

	case NAME:
	    strcpy(dirname, fname+n0);
	    state = FINISHED;
	    break;

	case FINISHED: // To keep the compiler happy
	    break;
	} /* switch */
    }

/* tail end validation */
    dp = getnodebyname(node);
    if (!dp)
    {
	lasterror = "Unknown or invalid node name ";
	return FALSE;
    }

    /* Default to a full wildcard if no filename is given. This will cope with
       the situation where no directory is provided (node::) and where a
       directory (node::[christine]) is provided. Logical names must end with
       a colon for this to happen or we will not be able to distinguish it
       from a filename

	EMS: Changed wildcard to *.*;0 so we only get the last version of
             the files. This is useful for the dapfs.
    lastchar = dirname[strlen(dirname)-1];
    if (lastchar == ']' || lastchar == ':' || dirname[0] == '\0')
	strcat(dirname, "*.*;0");

    */

    // Try very hard to get the local username
    local_user = cuserid(NULL);
    if (!local_user || local_user == (char *)0xffffffff)
      local_user = getenv("LOGNAME");
    if (!local_user) local_user = getenv("USER");
    if (local_user)
    {

	strcpy((char *)accessdata.acc_acc, local_user);
	accessdata.acc_accl = strlen((char *)accessdata.acc_acc);
        for (i=0; i<accessdata.acc_accl; i++)
            accessdata.acc_acc[i] = toupper(accessdata.acc_acc[i]);
    }
    else
        accessdata.acc_acc[0] = '\0';


    /* Complete the accessdata structure */
    accessdata.acc_userl = strlen(accessdata.acc_user);
    accessdata.acc_passl = strlen(accessdata.acc_pass);
    accessdata.acc_accl  = strlen(accessdata.acc_acc);
    return TRUE;
}

/*-----------------------------------------------------------------------*/
static int	dap_setup_link(void)
{
    int	sockfd;
    if ((sockfd=socket(AF_DECnet,SOCK_STREAM,DNPROTO_NSP)) == -1)
    {
	perror("socket");
	exit(-1);
    }


    if (setsockopt(sockfd,DNPROTO_NSP,SO_CONACCESS,&accessdata,
		   sizeof(accessdata)) < 0)
    {
	perror("setsockopt");
	exit(-1);
    }


    if (connect(sockfd, (struct sockaddr *)&sockaddr,
		sizeof(sockaddr)) < 0)
    {
	perror("socket");
	exit(-1);
    }

    return sockfd;
}
/*-------------------------------------------------------------------------*/
static void	dap_exchg_config(int sockfd)
{
    int		er;
    unsigned char	buf[100];
    unsigned char	confmsg[17] = {
	0x01,	/* Message type 1 = Conf */
	0x00,	/* Flag Field		 */
	0x2E,0x04, /* BufSiz 		 */
	0x07,	/* OS type = VMS	 */
	0x03,   /* RMS-32		 */
	7,0,0,0,0, /* Version 'stuff'     */
	0xA2,	/* Sequential Org Support*/
	0x80,
	0xB0,   /* Message Blocking. OK  */
	0xFE,   /* Supp Extended Attrib  */
	0xA8,   /* Supp. Seq Access/Del  */
	0x7C	/* WildCrds/Ren/Seg Msgs */
    };

    if ((er=write(sockfd,confmsg,sizeof(confmsg))) < 0)
    {
	perror("Send DAP config msg");
	exit(-1);
    }
    if ((er=read(sockfd,buf,sizeof(buf))) < 0)
    {
	perror("Recv DAP config msg");
	exit(-1);
    }
    /* Assume that Peer Supports the minimal characteristis we
       setup here
    */
    if (buf[0] == 0x09)
    {
	printf("Local DAP configuration rejected by remote node\n");
	exit(-1);
    }
}
/*-----------------------------------------------------------------------*/
/*-----------------------------------------------------------------------*/
/*-----------------------------------------------------------------------*/
void
dn_setmountentry(void)
{
	struct	mntent	mount_entry;
	char	vms_mount_point[250];
	FILE	*mtab;
	int	MFD;

	strcpy(vms_mount_point,node);
	strcat(vms_mount_point,"::");
	strcat(vms_mount_point,dirname);

	mount_entry.mnt_fsname = vms_mount_point;
	mount_entry.mnt_dir    = mount_point;
	mount_entry.mnt_type   = "dapfs";
	mount_entry.mnt_opts   = "rw";
	mount_entry.mnt_freq   = 0;
	mount_entry.mnt_passno = 0;



	if ( (MFD = open (MOUNTED "~", O_RDWR | O_CREAT | O_EXCL, 0600)) < 0)
	{
		perror("opening /etc/mtab file");
		exit(-1);
	}
	close(MFD);
	if ( (mtab = setmntent(MOUNTED, "a+")) == NULL)
	{
		perror("setmntent");
		exit(-1);
	}
	if ( addmntent(mtab,&mount_entry) == 1)
	{
		perror("addmntent");
		exit(-1);
	}
	if ( fchmod(fileno(mtab),0644) < 0)
	{
		perror ("fchmod");
		exit(-1);
	}
	endmntent(mtab);

	if ( unlink(MOUNTED "~") < 0 )
	{
		perror("unlink");
		exit(-1);
	}

}
/*-----------------------------------------------------------------------*/
/*-----------------------------------------------------------------------*/
int main(int argc, char *argv[])
{
	struct	stat	st;
	struct	passwd	*pwd;
	struct	group	*grp;
	char		opt;

	memset(&vms_directory,0,sizeof(vms_directory));
	memset(&mount_point,0,sizeof(mount_point));
	data.uid=getuid();
	data.gid=getgid();
	opterr=0;
	optind=0;
	while ( (opt=getopt(argc,argv,"?hu:g:")) != EOF)
	{
	  switch (opt)
	  {
		case '?':
		case 'h':
			  usage();
			  exit(0);
		case 'u':
			  if (isdigit(optarg[0]))
				data.uid=atoi(optarg);
			  else
			  {
				if ( (pwd=getpwnam(optarg)) == NULL)
				{
					printf("Invalid uid %s\n",optarg);
					exit (-1);
				}
				data.uid=pwd->pw_uid;
			  }
			  break;
		case 'g':
			  if (isdigit(optarg[0]))
				data.gid=atoi(optarg);
			  else
			  {
				if ( (grp=getgrnam(optarg)) == NULL)
				{
					printf("Invalid gid %s\n",optarg);
					exit (-1);
				}
				data.gid=grp->gr_gid;
			  }
			  break;
		default:  usage();
			  exit(0);
	  }
	}
	if ( (argc - optind)  < 2)
	{
		usage();
		exit(0);
	}
	memcpy(vms_directory,argv[optind],strlen((char *)argv[optind]));
	optind +=1;
	memcpy(mount_point,argv[optind],strlen((char *)argv[optind]));

	/* Parse the filename into its parts */
	if (!parse(vms_directory))
	{
		fprintf(stderr, "%s\n", lasterror);
		exit(2);
	}
	if (stat(mount_point,&st) < 0)
	{
		fprintf(stderr, "Invalid mount point %s : %s\n", mount_point,
				strerror(errno));
		exit(2);
	}
	if (!S_ISDIR(st.st_mode))
	{
		fprintf(stderr, "Mount point %s is not a directory\n",
				mount_point);
		exit(2);
	}
	if ( (getuid() != 0) && ((getuid() != st.st_uid) ||
	     ((st.st_mode & S_IRWXU) != S_IRWXU)))
	{
		fprintf(stderr, "Must be root to mount\n");
		exit(2);
	}
	strcpy(data.mounted_dir,dirname);
    	sockaddr.sdn_family = AF_DECnet;
    	sockaddr.sdn_flags	= 0x00;
    	sockaddr.sdn_objnum	= 0x11;			/* FAL		*/
    	sockaddr.sdn_objnamel	= 0x00;

    	memcpy(sockaddr.sdn_add.a_addr, dp->n_addr,2);
	memcpy(&data.sockaddr,&sockaddr,sizeof(sockaddr));
	memcpy(&data.accessdata,&accessdata,sizeof(accessdata));

	data.sockfd=dap_setup_link();
	dap_exchg_config(data.sockfd);
	if (mount(vms_directory,mount_point,
                  "dapfs",MS_MGC_VAL,(char *)&data) < 0)
	{
		fprintf(stderr, "Error mounting %s \n",vms_directory);
		exit(-1);
	}
	dn_setmountentry();
	return 0;
}
/*-------------------------------------------------------------------------*/
